eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxWits.hs

```diff
@@ -480,20 +497,41 @@ instance
   (Era era) =>
   FromCBOR (Annotator (RedeemersRaw era))
   where
-  fromCBOR = fmap RedeemersRaw <$> dec
+  fromCBOR =
+    fmap RedeemersRaw
+      <$> ifDecoderVersionAtLeast
+        (natVersion @9)
+        (dec decodeConway)
+        (dec decodePreConway)
     where
-      dec :: forall s. Decoder s (Annotator (Map RdmrPtr (Data era, ExUnits)))
-      dec = do
-        entries <- fmap sequence
-          . decodeList
-          . decodeRecordNamed "redeemer" (const 4)
-          $ do
-            rdmrPtr <- fromCBORGroup
-            dat <- fromCBOR
-            ex <- fromCBOR
-            let f x y z = (x, (y, z))
-            pure $ f rdmrPtr <$> dat <*> pure ex
-        pure $ Map.fromList <$> entries
+      dec :: Decoder s (Annotator (RdmrPtr, (Data era, ExUnits))) -> Decoder s (Annotator (Map RdmrPtr (Data era, ExUnits)))
+      dec decoder = fmap (fmap Map.fromList) (fmap sequence . decodeList $ decoder)
+      decodeRedeemerRecord :: Decoder s (Annotator (RdmrPtr, (Data era, ExUnits)))
+      decodeRedeemerRecord = do
+        rdmrPtr <- fromCBORGroup
+        dataEra <- fromCBOR
+        exUnits <- fromCBOR
+        let f :: RdmrPtr -> (Data era) -> ExUnits -> (RdmrPtr, (Data era, ExUnits))
+            f x y z = (x, (y, z))
+        pure $ f rdmrPtr <$> dataEra <*> pure exUnits
+      decodePreConway :: Decoder s (Annotator (RdmrPtr, (Data era, ExUnits)))
+      decodePreConway = decodeRecordNamed "redeemer" (const 4) $ decodeRedeemerRecord
+      -- This implementation is adapted directly from `decodeRecordNamed`
+      -- It is customized for Conway because in this Era, we want to allow the decoding of
+      -- Redeemer that is encoded using both sizes, the previous, wrong size of 4, and the new, correct size of 3.
+      decodeConway :: Decoder s (Annotator (RdmrPtr, (Data era, ExUnits)))
+      decodeConway = runIdentityT $ do
+        lenOrIndef <- lift decodeListLenOrIndef
+        record <- lift decodeRedeemerRecord
+        lift $ case lenOrIndef of
+          Just n ->
+            when (n `notElem` [3, 4]) $
+              cborError $
+                DecoderErrorSizeMismatch (Text.pack "Record redeemers should be sized 3 OR 4 for Conway") n n
+          Nothing -> do
+            isBreak <- decodeBreakOr
+            unless isBreak $ cborError $ DecoderErrorCustom "redeemers" "Excess terms in array"
+        pure record
```

Comment:
Identity transformer wrapper is not need. Also we don't need to report 4 as an option in the error message. 3 is the correct value, 4 is merely a compatibility check.
```suggestion
      decodeConway = do
        lenOrIndef <- decodeListLenOrIndef
        record <- decodeRedeemerRecord
        case lenOrIndef of
          Just n ->
            when (n `notElem` [3, 4]) $
              cborError $
                DecoderErrorSizeMismatch (Text.pack "Record redeemers should be sized 3 for Conway") 3 n
          Nothing -> do
            isBreak <- decodeBreakOr
            unless isBreak $ cborError $ DecoderErrorCustom "redeemers" "Excess terms in array"
        pure record
```